File: python/robot-simulator/robot_simulator.py
# Globals for the directions
# Change the values as you see fit
EAST = 1
NORTH = 0
WEST = 3
SOUTH = 2


# Robot class definition
class Robot:
    # Constructor to initialize the robot's direction and coordinates
    def __init__(self, direction=NORTH, x=0, y=0):
        self.coordinates = (x, y)  # Initialize coordinates as a tuple (x, y)
        self.direction = direction  # Initialize direction (NORTH by default)

    # Method to turn the robot right
    def turn_right(self):
        self.direction = (self.direction + 1) % 4  # Update direction clockwise

    # Method to turn the robot left
    def turn_left(self):
        for _ in range(3):
            self.turn_right()  # Equivalent to turning right three times (270-degree left turn)

    # Method to move the robot forward one step in its current direction
    def advance(self):
        x, y = self.coordinates  # Get current coordinates (x, y)

        # Update coordinates based on the current direction
        if self.direction == NORTH:
            self.coordinates = (x, y + 1)  # Move one step north (increment y)
        elif self.direction == EAST:
            self.coordinates = (x + 1, y)  # Move one step east (increment x)
        elif self.direction == SOUTH:
            self.coordinates = (x, y - 1)  # Move one step south (decrement y)
        elif self.direction == WEST:
            self.coordinates = (x - 1, y)  # Move one step west (decrement x)

    # Method to execute a sequence of instructions (R: turn right, L: turn left, A: advance)
    def move(self, instructions):
        for ins in instructions:
            if ins == 'R':
                self.turn_right()  # Turn right
            elif ins == 'L':
                self.turn_left()  # Turn left (270-degree left turn)
            elif ins == 'A':
                self.advance()  # Move one step forward

File: javascript/robot-simulator/robot-simulator.js
export class InvalidInputError extends Error {
  constructor(message) {
    super();
    this.message = message || "Invalid Input";
  }
}

export class Robot {
  // Constants for cardinal directions and instructions
  static NORTH = "north";
  static EAST = "east";
  static SOUTH = "south";
  static WEST = "west";
  static BEARINGS = [Robot.NORTH, Robot.EAST, Robot.SOUTH, Robot.WEST];
  static INSTRUCTIONS = {
    A: "advance",
    L: "turnLeft",
    R: "turnRight",
  };

  // Static method to convert chain of instructions to instruction names
  static instructions(chain) {
    return [...chain].map((instruction) => Robot.INSTRUCTIONS[instruction]);
  }

  // Private properties
  #bearing = Robot.NORTH;
  #coordinates;

  // Getter for bearing
  get bearing() {
    return this.#bearing;
  }

  // Getter for coordinates
  get coordinates() {
    return [...this.#coordinates];
  }

  // Set the robot's orientation (bearing)
  orient(bearing) {
    // Throw error for invalid bearing
    if (!Robot.BEARINGS.includes(bearing)) throw new InvalidInputError();
    this.#bearing = bearing;
  }

  // Set the robot's initial coordinates
  at(...coordinates) {
    this.#coordinates = coordinates;
  }

  // Turn the robot right based on current orientation
  turnRight() {
    this.orient(
      Robot.BEARINGS[
        (Robot.BEARINGS.indexOf(this.#bearing) + 1) % Robot.BEARINGS.length
      ],
    );
  }

  // Turn the robot left based on current orientation
  turnLeft() {
    this.orient(
      Robot.BEARINGS[
        (Robot.BEARINGS.indexOf(this.#bearing) + 3) % Robot.BEARINGS.length
      ],
    );
  }

  // Move the robot forward based on its orientation
  advance() {
    switch (this.bearing) {
      case Robot.NORTH:
        return this.#coordinates[1]++;
      case Robot.EAST:
        return this.#coordinates[0]++;
      case Robot.SOUTH:
        return this.#coordinates[1]--;
      case Robot.WEST:
        return this.#coordinates[0]--;
    }
  }

  // Place the robot at a specific position and orientation
  place({ x, y, direction }) {
    this.orient(direction);
    this.at(x, y);
  }

  // Evaluate a chain of instructions
  evaluate(chain) {
    Robot.instructions(chain).forEach((instruction) => this[instruction]());
  }
}
